contracts/Smt.metasol

```diff
@@ -0,0 +1,139 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.7.0;
+
+library SMT{{hashname}} {
+    // because maxElements = 2**depth, so 256 will overflow
+    uint8 constant MAX_DEPTH = 255;
+    bytes32 constant internal EMPTY_SUBTREE = 0x00;
+
+    struct MerkleTree {
+        uint8 depth;
+        mapping(uint256 => bytes32) cacheEmptyValues;
+        // tree level => index inside level => element hash
+        mapping(uint256 => mapping(uint256 => bytes32)) tree;
+        mapping(uint256 => bytes) elementData;
+    }
+
+    using SMT{{hashname}} for MerkleTree;
+
+    function initialize(MerkleTree storage tree, uint8 _depth) internal {
+        require(_depth > 0, "Depth must be non-zero");
+        tree.cacheEmptyValues[0] = {{hash}}("");
+        // current depth == 0
+        tree._calculateEmptyLeafHash(_depth);
+        tree.depth = _depth;
+    }
+
+    function totalElements(MerkleTree storage tree) internal view returns (uint256) {
+        // 1<<depth == 2**depth
+        return 1<<tree.depth;
+    }
+
+    function increaseDepth(MerkleTree storage tree, uint8 depthDifference) internal {
```

Comment:
Should check that `tree.depth > 0`

Reply:
It is not safe to increase the size of a tree in which there are already values. Should check the same as in `decreaseDepth`

Reply:
Why isn't it safe? our old tree just becomes a subtree in the new tree with increased depth.

Reply:
Sorry, I drew a tree with different indexing and was sure that there would be problems, but right now I re-read the code and realized that everything is ok 🙃

Reply:
BTW, note that in your drawing the level numbering is reversed: leaves are (should be) level 0 and root is level _depth_.

Reply:
We should check that `tree.depth` > 0 anyway